src/psyclone/psyir/transformations/omp_canonical_loop_trans.py

```diff
@@ -0,0 +1,254 @@
+        if not isinstance(node, BinaryOperation):
+            return False
+        # We allow var + a2, a2 + var
+        if (node.operator == BinaryOperation.Operator.ADD and
+                isinstance(node.children[0], Reference) and
+                isinstance(node.children[1], Reference)):
+            return True
+
+        # We allow var - a2 and a2 - var
+        if (node.operator == BinaryOperation.Operator.SUB and
+                isinstance(node.children[0], Reference) and
+                isinstance(node.children[1], Reference)):
+            return True
+
+        # We allow var * a1 and a1 * var
+        if (node.operator == BinaryOperation.Operator.MUL and
+                isinstance(node.children[0], Reference) and
+                isinstance(node.children[1], Reference)):
+            return True
+
+        # Simple cases analysed, remaining allowed cases:
+        # a1 * var + a2  - Handled
+        # a2 + a1 * var - Handled
+        # a1 * var - a2
+        # a2 - a1 * var
+        # var * a1 + a2 - Handled
+        # a2 + var * a1 - Handled
+        # var * a1 - a2
+        # a2 - var * a1
+
+        # Check for a1 * var + a2 - temporary version. Quick test
+        # seems to show this is
+        # BinaryOperation[ADD:BinaryOperation[MUL:a1, var],a2]
+        # We allow the inner binary operation to be a1*var or var*a1
+        # I think this also handles a2 + (a1 * var)
+        if (node.operator == BinaryOperation.Operator.ADD and
+                isinstance(node.children[1], Reference) and
+                node.children[1].symbol != loop_var and
+                isinstance(node.children[0], BinaryOperation) and
+                node.children[1].operator == BinaryOperation.Operator.MUL and
+                (node.children[0].children[0].symbol == loop_var or
+                 node.children[0].children[1].symbol == loop_var)):
+            return True
+
+        # Check for a2 - a1*var or a2 - var*a1
+        if (node.operator == BinaryOperation.Operator.SUB and
+                isinstance(node.children[0], Reference) and
+                node.children[0].symbol != loop_var and
+                isinstance(node.children[1], BinaryOperation) and
+                node.children[1].operator == BinaryOperation.Operator.MUL and
+                (node.children[1].children[0].symbol == loop_var or
+                 node.children[1].children[1].symbol == loop_var)):
+            return True
+
+        # Check for a1*var - a2 or var*a1 - a2
+        if (node.operator == BinaryOperation.Operator.SUB and
+                isinstance(node.children[1], Reference) and
+                node.children[1].symbol != loop_var and
+                isinstance(node.children[0], BinaryOperation) and
+                node.children[0].operator == BinaryOperation.Operator.MUL and
+                (node.children[0].children[0].symbol == loop_var or
+                 node.children[0].children[1].symbol == loop_var)):
+            return True
```

Comment:
These tests still don't feel right. Too many very similar tests, and I am not even sure that they are complete (what about -a1*var-a2 or +a1*var+a2, which from memory adds another UnaryOperation node).
SymPy offers a polynomial class, which gives you access to the coefficients, and for the example above at least it will convert them all to a similar structure:
```
>>> sp.Poly(a1 * var - a2, var).all_coeffs()
[a1, -a2]
>>> sp.Poly(a2 - a1 * var,var).all_coeffs()
[-a1, a2]
```
etc. Unfortunately, it does still first simplify the expressions, so
```
>>> sp.Poly(a2 - a1 * var + 2*var,var).all_coeffs()
[2 - a1, a2]
```
Could we perhaps in a case like this rewrite the PSyIR with the equivalent expression (psyad is already doing this when expanding terms). Or you could additionally count the number of references and symbols in the original PSyIR, which should be less than or equal to 4?

Reply:
I looked again and my interpretation of the OpenMP standard was also wrong, in that `a1` and `a2` simply need to be "loop invariant integer expressions", and not simply variables.
I think I should have already checked that any `Reference` objects in these terms are loop invariant in `validate` before reaching this function, so what actually needs to be done is:
1. Check we have some structure that could be handled as one of the allowed structures.
2. Check that the `a1` and `a2` expressions (if present) are integer expressions.
I wonder if 2 would have been handled already during input/tree construction or if we can just ignore it.
I don't have a good idea how to check the structure conforms though, maybe the sp.Poly class you recommend will work straight out the box? But I'm not clear how to use this - could you provide a quick example?

I'm trying to work out what this actually disallows, and I think any operation involving all integer values/references, and only `+-*` operations is allowed (since you could transform `x*y*var_outer` into `a1*var_outer` since `a1=x*y`. Equally division (maybe any integer-safe) operations should be ok provided they are not performed on `var_outer`. Does that make sense or have I missed something?